filer/models/filemodels.py

```diff
@@ -432,24 +432,29 @@ def delete(self, *args, **kwargs):
         super(File, self).delete_restorable(*args, **kwargs)
     delete.alters_data = True
 
-    def _generate_valid_location_for_restore(self):
+    def _generate_valid_name_for_restore(self):
         """
-        Returns the first available destination path where this file
-            will be restored to.
+        Generates the first available name so this file
+            can be restored in the folder.
         """
-        basename, extension = os.path.splitext(self.actual_name)
-        destination = self.file.field.upload_to(self, self.actual_name)
+        basename, extension = os.path.splitext(self.clean_actual_name)
+        if self.folder:
+            files = self.folder.files
+        elif self.owner:
+            files = filer.models.tools.get_user_clipboard(self.owner).files.all()
+        else:
+            from filer.models.virtualitems import UnfiledImages
```

Comment:
https://github.com/pbs/django-filer/blob/6c1b462eb48d4aa5add4a670b0edfb65d2322807/filer/models/filemodels.py#L580
It's a circular import, that is already handled in this way so I've kept it this way.